[docs] MongoDB auto-user provisioning (#34939)
gravitational/teleport: docs/pages/database-access/auto-user-provisioning/mongodb.mdx

---
title: MongoDB Automatic User Provisioning 
description: Configure automatic user provisioning for MongoDB.
---

(!docs/pages/includes/database-access/auto-user-provisioning/intro.mdx!)

> - Teleport cluster v14.3 or above with a configured [self-hosted
>   MongoDB](../guides/mongodb-self-hosted.mdx).

Reviewer: ```suggestion
- Teleport cluster v14.3 or above.
- A self-hosted
  MongoDB database enrolled with your Teleport cluster. Follow the [Teleport documentation](../guides/mongodb-self-hosted.mdx) to learn how to enroll your database.
```
Suggested to be more explicit about what "with a configured..." means and what the link points to.

- Ability to connect to and create user accounts in the target database.

<Admonition type="note" title="Supported services">
Automatic user provisioning is not compatible with MongoDB Altas.
</Admonition>

## Step 1/3. Configure database admin

Teleport uses the same authentication mechanism (X.509) when connecting as an
admin user as for regular user connections.

The admin user must have privileges within the database to create users and
grant them privileges. The admin user must also have privileges to monitor user
connections.

> First create a role on `admin` database with the following privileges:
> ```bash

Reviewer: ```suggestion
First, create a role on the `admin` database with the following privileges:

```bash
```

db.getSiblingDB("admin").runCommand({
    createRole: "teleport-admin-role",
    privileges: [
        { resource: { cluster: true }, actions: [ "inprog" ] },
        { resource: { db: "", collection: "" }, actions: [ "grantRole", "revokeRole" ] },
        { resource: { db: "$external", "collection": "" }, actions: [ "createUser", "updateUser", "dropUser", "viewUser", "setAuthenticationRestriction", "changeCustomData"] },
    ],
    roles: [],
})
```

> <Details title="Limit grantRole action to specific databases">

Reviewer: ```suggestion
<Details title="Limit the grantRole action to specific databases">
```

In the above example, the `grantRole` privilege is granted to the admin user
for all databases in order for the admin user to assign roles from all
databases, including the `admin` database.

To enforce the principle of least privilege, you can limit the `grantRole` to
only the databases that own the roles to be assigned to the auto-provisioned
users:

```bash
db.getSiblingDB("admin").runCommand({
    createRole: "teleport-admin-role",
    privileges: [
        { resource: { cluster: true }, actions: [ "inprog" ] },
        { resource: { db: "", collection: "" }, actions: [ "revokeRole" ] },
        { resource: { db: "$external", "collection": "" }, actions: [ "createUser", "updateUser", "dropUser", "viewUser", "setAuthenticationRestriction", "changeCustomData"] },
        { resource: { db: "<db1>", collection: "" }, actions: [ "grantRole" ] },
        { resource: { db: "<db2>", collection: "" }, actions: [ "grantRole" ] },
        ...
    ],
    roles: [],
})
```
</Details>

> Now create the admin user with this role:

Reviewer: ```suggestion
Now create the admin user with this role:

```

```bash
db.getSiblingDB("$external").runCommand({
  createUser: "CN=teleport-admin",
  roles: [ { role: 'teleport-admin-role', db: 'admin' } ],
})
```

(!docs/pages/includes/database-access/auto-user-provisioning/db-definition-self-hosted.mdx protocol="mongodb" uri="localhost:27017"!)

> ## Step 2/3. Configure Teleport role

Reviewer: ```suggestion
## Step 2/3. Configure a Teleport role
```

To specify the database roles a user should be assigned within the database,
use the `db_roles` role option:

```yaml
kind: role
version: v7
metadata:
  name: auto-db-users
spec:
  options:
    # create_db_user_mode enables automatic user provisioning for matching databases
    create_db_user_mode: keep
  allow:
    db_labels:
      "*": "*"
    db_names:
    - "*"
    # db_roles is a list of roles the database user will be assigned
    db_roles:
    - "readAnyDatabase@admin"
    - "readWrite@db1"
    - "myCustomRole@db2"
    - "{{internal.db_roles}}"
    - "{{external.db_roles}}"
```

(!docs/pages/includes/database-access/auto-user-provisioning/db_users_ignored.mdx!)

(!docs/pages/includes/database-access/auto-user-provisioning/modes.mdx!)

Users created within the database will:

> - Have the same name as Teleport username.

Reviewer: ```suggestion
- Have the same username as the authenticated Teleport user.
```

> - Have `teleport-auto-user` set to `true` in user's `customData`.

Reviewer: ```suggestion
- Have `teleport-auto-user` set to `true` in the user's `customData`.
```
